File: Banco de Dados/import_planilha.py
import argparse
import os
import re
import sqlite3
import zipfile
import xml.etree.ElementTree as ET
from datetime import date, datetime, timedelta

NS = "{http://schemas.openxmlformats.org/spreadsheetml/2006/main}"
R_NS = "{http://schemas.openxmlformats.org/officeDocument/2006/relationships}"
PKG_REL_NS = "{http://schemas.openxmlformats.org/package/2006/relationships}"


def excel_date(value):
    if value in (None, ""):
        return None
    try:
        # Excel serial date. 1899-12-30 handles Excel's leap-year bug.
        return (datetime(1899, 12, 30) + timedelta(days=float(value))).date().isoformat()
    except Exception:
        return None


def to_float(value):
    if value in (None, ""):
        return 0.0
    if isinstance(value, (int, float)):
        return float(value)
    txt = str(value).strip()
    if not txt or txt == "-":
        return 0.0
    txt = txt.replace("R$", "").replace("%", "").strip()
    # Formato brasileiro: 1.234,56
    if "," in txt and "." in txt:
        txt = txt.replace(".", "").replace(",", ".")
    elif "," in txt:
        txt = txt.replace(",", ".")
    try:
        return float(txt)
    except Exception:
        return 0.0


def to_int_from_text(value):
    if value in (None, ""):
        return 0
    m = re.search(r"\d+", str(value))
    return int(m.group(0)) if m else 0


def split_product_and_sku(text):
    if not text:
        return "", ""
    parts = str(text).splitlines()
    nome = parts[0].strip() if parts else str(text).strip()
    sku = ""
    for part in parts[1:]:
        # Ex.: "M004068" ou "Cor: Verde | SKU M005944T"
        m = re.search(r"(?:SKU\s*)?([A-Z]\d{3,}[A-Z0-9]*)", part.strip(), re.I)
        if m:
            sku = m.group(1).strip()
            break
    return nome, sku


def calc_promotion_status(data_inicio, data_fim, today=None):
    today = today or date.today()
    if not data_inicio and not data_fim:
        return "Sem Promoção"
    try:
        inicio = date.fromisoformat(data_inicio) if data_inicio else None
        fim = date.fromisoformat(data_fim) if data_fim else None
    except Exception:
        return "Sem Promoção"
    if inicio and inicio > today:
        return "Programado"
    if fim:
        if fim < today:
            return "Expirado"
        return "Ativado"
    return "Ativado"


class XlsmReader:
    def __init__(self, path):
        self.path = path
        self.zip = zipfile.ZipFile(path)
        self.shared_strings = self._load_shared_strings()
        self.sheets = self._load_sheet_map()

    def _load_shared_strings(self):
        strings = []
        try:
            root = ET.fromstring(self.zip.read("xl/sharedStrings.xml"))
        except KeyError:
            return strings
        for si in root.findall(NS + "si"):
            texts = []
            for t in si.iter(NS + "t"):
                texts.append(t.text or "")
            strings.append("".join(texts))
        return strings

    def _load_sheet_map(self):
        wb = ET.fromstring(self.zip.read("xl/workbook.xml"))
        rels = ET.fromstring(self.zip.read("xl/_rels/workbook.xml.rels"))
        rel_map = {rel.attrib["Id"]: rel.attrib["Target"] for rel in rels}
        sheets = {}
        for sh in wb.find(NS + "sheets"):
            name = sh.attrib.get("name")
            rid = sh.attrib.get(R_NS + "id")
            target = rel_map.get(rid)
            if target:
                sheets[name] = target.replace("worksheets/", "")
        return sheets

    @staticmethod
    def _cell_ref_to_pos(ref):
        m = re.match(r"([A-Z]+)(\d+)", ref)
        col = 0
        for ch in m.group(1):
            col = col * 26 + ord(ch) - 64
        return int(m.group(2)), col

    def get_hyperlinks(self, sheet_xml_name):
        rel_path = f"xl/worksheets/_rels/{sheet_xml_name}.rels"
        rels = {}
        try:
            rel_root = ET.fromstring(self.zip.read(rel_path))
            rels = {rel.attrib["Id"]: rel.attrib.get("Target", "") for rel in rel_root}
        except KeyError:
            pass
        root = ET.fromstring(self.zip.read(f"xl/worksheets/{sheet_xml_name}"))
        links = {}
        for h in root.iter(NS + "hyperlink"):
            ref = h.attrib.get("ref")
            rid = h.attrib.get(R_NS + "id")
            if ref and rid in rels:
                links[ref] = rels[rid]
        return links

    def sheet_rows(self, sheet_name):
        sheet_xml = self.sheets[sheet_name]
        root = ET.fromstring(self.zip.read(f"xl/worksheets/{sheet_xml}"))
        cells = {}
        for c in root.iter(NS + "c"):
            ref = c.attrib.get("r")
            row, col = self._cell_ref_to_pos(ref)
            cell_type = c.attrib.get("t")
            v = c.find(NS + "v")
            inline = c.find(NS + "is")
            value = ""
            if cell_type == "s" and v is not None:
                value = self.shared_strings[int(v.text)]
            elif cell_type == "inlineStr" and inline is not None:
                value = "".join(t.text or "" for t in inline.iter(NS + "t"))
            elif v is not None:
                value = v.text
            else:
                f = c.find(NS + "f")
                if f is not None:
                    value = "=" + (f.text or "")
            cells[(row, col)] = value
        max_row = max((r for r, c in cells), default=0)
        max_col = max((c for r, c in cells), default=0)
        rows = []
        for r in range(1, max_row + 1):
            rows.append([cells.get((r, c), "") for c in range(1, max_col + 1)])
        return rows


def import_to_db(planilha, db_path, schema_path):
    reader = XlsmReader(planilha)
    conn = sqlite3.connect(db_path)
    conn.execute("PRAGMA foreign_keys = ON")
    with open(schema_path, "r", encoding="utf-8") as f:
        conn.executescript(f.read())

    # Limpa dados operacionais para reimportação completa.
    for table in ["simulacoes", "vendas", "promocoes", "anuncios", "importacoes"]:
        conn.execute(f"DELETE FROM {table}")
    conn.commit()

    anuncios_count = 0
    promocoes_count = 0
    vendas_count = 0

    # Anúncios / Precificação
    prec_rows = reader.sheet_rows("Precificação de produto")
    prec_links = reader.get_hyperlinks(reader.sheets["Precificação de produto"])
    for idx, row in enumerate(prec_rows[3:], start=4):
        if len(row) < 13:
            continue
        codigo = str(row[2]).strip()
        nome = str(row[4]).strip()
        if not codigo or not nome or codigo == "Cód. do Produto":
            continue
        link = prec_links.get(f"A{idx}", "")
        tipo = "Premium" if to_float(row[7]) >= 0.16 else "Clássico"
        conn.execute(
            """
            INSERT OR REPLACE INTO anuncios
            (codigo_produto, sku, nome, status, tipo_anuncio, valor_total, frete, tarifa,
             custo_fixo, desconto, preco_anunciado, liquido_receber, estoque,
             quantidade_vendida, ultima_atualizacao_preco, link_anuncio, ordem_importacao)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            """,
            (
                codigo,
                str(row[3]).strip(),
                nome,
                str(row[1]).strip(),
                tipo,
                to_float(row[5]),
                to_float(row[6]),
                to_float(row[7]),
                to_float(row[8]),
                to_float(row[9]),
                to_float(row[10]),
                to_float(row[11]),
                0,
                to_int_from_text(row[12]),
                None,
                link,
                anuncios_count + 1,
            ),
        )
        anuncios_count += 1

    # Promoções
    promo_rows = reader.sheet_rows("Central de Promoções")
    promo_links = reader.get_hyperlinks(reader.sheets["Central de Promoções"])
    for idx, row in enumerate(promo_rows[3:], start=4):
        if len(row) < 9:
            continue
        codigo = str(row[2]).strip()
        nome = str(row[3]).strip()
        if not codigo or not nome or codigo == "Cód. do Produto":
            continue
        data_inicio = excel_date(row[4])
        data_fim = excel_date(row[5])
        status_raw = str(row[0]).strip()
        status = status_raw if status_raw and not status_raw.startswith("=") else calc_promotion_status(data_inicio, data_fim)
        if not data_inicio and not data_fim:
            status = "Sem Promoção"
        anuncio = conn.execute("SELECT id FROM anuncios WHERE codigo_produto = ?", (codigo,)).fetchone()
        conn.execute(
            """
            INSERT INTO promocoes
            (anuncio_id, codigo_produto, nome_anuncio, status, data_inicio, data_fim,
             preco, desconto, preco_final, link_promocao)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            """,
            (
                anuncio[0] if anuncio else None,
                codigo,
                nome,
                status,
                data_inicio,
                data_fim,
                to_float(row[6]),
                to_float(row[7]),
                to_float(row[8]),
                promo_links.get(f"B{idx}", ""),
            ),
        )
        promocoes_count += 1

    # Vendas
    vendas_rows = reader.sheet_rows("Vendas")
    vendas_links = reader.get_hyperlinks(reader.sheets["Vendas"])
    for idx, row in enumerate(vendas_rows[2:], start=3):
        if len(row) < 10:
            continue
        numero = str(row[0]).strip()
        if not numero or numero == "Nº Venda":
            continue
        produto_nome, sku_extraido = split_product_and_sku(row[3])
        codigo = str(row[4]).strip()
        anuncio = conn.execute("SELECT id, sku FROM anuncios WHERE codigo_produto = ?", (codigo,)).fetchone()
        sku = sku_extraido or (anuncio[1] if anuncio else "")
        conn.execute(
            """
            INSERT OR REPLACE INTO vendas
            (numero_venda, anuncio_id, codigo_produto, sku, produto_nome, cliente_nome,
             status, data_venda, quantidade, valor_unitario, valor_receber, observacao, link_venda)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            """,
            (
                numero,
                anuncio[0] if anuncio else None,
                codigo,
                sku,
                produto_nome,
                str(row[9]).strip(),
                str(row[1]).strip(),
                excel_date(row[2]),
                int(to_float(row[5]) or 0),
                to_float(row[6]),
                to_float(row[7]),
                str(row[8]).strip(),
                vendas_links.get(f"A{idx}", ""),
            ),
        )
        vendas_count += 1

    conn.execute(
        """
        INSERT INTO importacoes
        (arquivo_origem, anuncios_importados, promocoes_importadas, vendas_importadas, observacao)
        VALUES (?, ?, ?, ?, ?)
        """,
        (os.path.basename(planilha), anuncios_count, promocoes_count, vendas_count,
         "Importação inicial a partir da planilha. Campos sensíveis como CPF/endereço não foram importados no MVP."),
    )
    conn.commit()
    conn.close()
    return anuncios_count, promocoes_count, vendas_count


def main():
    parser = argparse.ArgumentParser(description="Importa planilha xlsm para SQLite")
    parser.add_argument("--planilha", default="Gestão de Anúncio - Mercado Livre(2).xlsm")
    parser.add_argument("--db", default="gestao_vendas.sqlite")
    parser.add_argument("--schema", default="schema.sql")
    args = parser.parse_args()
    counts = import_to_db(args.planilha, args.db, args.schema)
    print(f"Importação concluída: anúncios={counts[0]}, promoções={counts[1]}, vendas={counts[2]}")


if __name__ == "__main__":
    main()

File: scripts/import-planilha.js
const fs = require('fs');
const path = require('path');
const XLSX = require('xlsx');
const { DatabaseSync } = require('node:sqlite');

const rootDir = path.resolve(__dirname, '..');
const dataDir = path.join(rootDir, 'Banco de Dados');
const spreadsheetPath = process.argv[2]
  ? path.resolve(process.argv[2])
  : path.join(dataDir, 'Gestão de Anúncio - Mercado Livre(2).xlsm');
const dbPath = process.argv[3]
  ? path.resolve(process.argv[3])
  : path.join(dataDir, 'gestao_vendas.sqlite');
const schemaPath = path.join(dataDir, 'schema.sql');

function toFloat(value) {
  if (value === null || value === undefined || value === '') return 0;
  if (typeof value === 'number') return value;
  const text = String(value).replace('R$', '').replace('%', '').trim();
  if (!text || text === '-') return 0;
  const normalized = text.includes(',') && text.includes('.')
    ? text.replace(/\./g, '').replace(',', '.')
    : text.replace(',', '.');
  const parsed = Number(normalized);
  return Number.isFinite(parsed) ? parsed : 0;
}

function toIntFromText(value) {
  const match = String(value ?? '').match(/\d+/);
  return match ? Number(match[0]) : 0;
}

function excelDate(value) {
  if (!value) return null;
  if (value instanceof Date && !Number.isNaN(value.getTime())) {
    return value.toISOString().slice(0, 10);
  }
  if (typeof value === 'number') {
    const date = XLSX.SSF.parse_date_code(value);
    if (!date) return null;
    return `${date.y}-${String(date.m).padStart(2, '0')}-${String(date.d).padStart(2, '0')}`;
  }
  const parsed = new Date(value);
  return Number.isNaN(parsed.getTime()) ? null : parsed.toISOString().slice(0, 10);
}

function splitProductAndSku(text) {
  const lines = String(text ?? '').split(/\r?\n/).map((line) => line.trim()).filter(Boolean);
  const nome = lines[0] || '';
  const skuLine = lines.find((line) => /SKU|[A-Z]\d{3,}/i.test(line)) || '';
  const sku = skuLine.match(/(?:SKU\s*)?([A-Z]\d{3,}[A-Z0-9]*)/i)?.[1] || '';
  return { nome, sku };
}

function calcPromotionStatus(dataInicio, dataFim) {
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  if (!dataInicio && !dataFim) return 'Sem Promoção';
  const inicio = dataInicio ? new Date(`${dataInicio}T00:00:00`) : null;
  const fim = dataFim ? new Date(`${dataFim}T00:00:00`) : null;
  if (inicio && inicio > today) return 'Programado';
  if (fim && fim < today) return 'Expirado';
  return 'Ativado';
}

function rows(workbook, sheetName) {
  const sheet = workbook.Sheets[sheetName];
  if (!sheet) throw new Error(`Aba não encontrada: ${sheetName}`);
  const data = [];
  for (const cellRef of Object.keys(sheet)) {
    if (cellRef.startsWith('!')) continue;
    const position = XLSX.utils.decode_cell(cellRef);
    if (!data[position.r]) data[position.r] = [];
    data[position.r][position.c] = sheet[cellRef].v ?? '';
  }
  return data.map((row) => row || []);
}

function link(workbook, sheetName, cellRef) {
  return workbook.Sheets[sheetName]?.[cellRef]?.l?.Target || '';
}

function importSpreadsheet() {
  if (!fs.existsSync(spreadsheetPath)) {
    throw new Error(`Planilha não encontrada: ${spreadsheetPath}`);
  }

  console.log(`Lendo planilha: ${spreadsheetPath}`);
  const workbook = XLSX.readFile(spreadsheetPath, { cellDates: true });
  console.log(`Abas encontradas: ${workbook.SheetNames.join(', ')}`);
  const db = new DatabaseSync(dbPath, { timeout: 1000 });
  db.exec('PRAGMA foreign_keys = ON');
  db.exec(fs.readFileSync(schemaPath, 'utf8'));

  console.log(`Recriando dados em: ${dbPath}`);
  db.exec('BEGIN');
  try {
    for (const table of ['simulacoes', 'vendas', 'promocoes', 'anuncios', 'importacoes']) {
      db.prepare(`DELETE FROM ${table}`).run();
    }

    const insertAnuncio = db.prepare(`
      INSERT OR REPLACE INTO anuncios
      (codigo_produto, sku, nome, status, tipo_anuncio, valor_total, frete, tarifa,
       custo_fixo, desconto, preco_anunciado, liquido_receber, estoque,
       quantidade_vendida, ultima_atualizacao_preco, link_anuncio, ordem_importacao)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    `);

    let anunciosCount = 0;
    const anunciosRows = rows(workbook, 'Precificação de produto').slice(3);
    console.log(`Linhas de anúncios lidas: ${anunciosRows.length}`);
    for (const [index, row] of anunciosRows.entries()) {
      const excelRow = index + 4;
      const codigo = String(row[2] ?? '').trim();
      const nome = String(row[4] ?? '').trim();
      if (!codigo || !nome || codigo === 'Cód. do Produto') continue;
      const tipo = toFloat(row[7]) >= 0.16 ? 'Premium' : 'Clássico';
      insertAnuncio.run(
        codigo,
        String(row[3] ?? '').trim(),
        nome,
        String(row[1] ?? '').trim(),
        tipo,
        toFloat(row[5]),
        toFloat(row[6]),
        toFloat(row[7]),
        toFloat(row[8]),
        toFloat(row[9]),
        toFloat(row[10]),
        toFloat(row[11]),
        0,
        toIntFromText(row[12]),
        null,
        link(workbook, 'Precificação de produto', `A${excelRow}`),
        anunciosCount + 1,
      );
      anunciosCount += 1;
    }

    const findAnuncio = db.prepare('SELECT id, sku FROM anuncios WHERE codigo_produto = ?');
    const insertPromocao = db.prepare(`
      INSERT INTO promocoes
      (anuncio_id, codigo_produto, nome_anuncio, status, data_inicio, data_fim,
       preco, desconto, preco_final, link_promocao)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    `);

    let promocoesCount = 0;
    const promocoesRows = rows(workbook, 'Central de Promoções').slice(3);
    console.log(`Linhas de promoções lidas: ${promocoesRows.length}`);
    for (const [index, row] of promocoesRows.entries()) {
      const excelRow = index + 4;
      const codigo = String(row[2] ?? '').trim();
      const nome = String(row[3] ?? '').trim();
      if (!codigo || !nome || codigo === 'Cód. do Produto') continue;
      const dataInicio = excelDate(row[4]);
      const dataFim = excelDate(row[5]);
      const statusRaw = String(row[0] ?? '').trim();
      const status = statusRaw && !statusRaw.startsWith('=')
        ? statusRaw
        : calcPromotionStatus(dataInicio, dataFim);
      const anuncio = findAnuncio.get(codigo);
      insertPromocao.run(
        anuncio?.id || null,
        codigo,
        nome,
        !dataInicio && !dataFim ? 'Sem Promoção' : status,
        dataInicio,
        dataFim,
        toFloat(row[6]),
        toFloat(row[7]),
        toFloat(row[8]),
        link(workbook, 'Central de Promoções', `B${excelRow}`),
      );
      promocoesCount += 1;
    }

    const insertVenda = db.prepare(`
      INSERT OR REPLACE INTO vendas
      (numero_venda, anuncio_id, codigo_produto, sku, produto_nome, cliente_nome,
       status, data_venda, quantidade, valor_unitario, valor_receber, observacao, link_venda)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    `);

    let vendasCount = 0;
    const vendasRows = rows(workbook, 'Vendas').slice(2);
    console.log(`Linhas de vendas lidas: ${vendasRows.length}`);
    for (const [index, row] of vendasRows.entries()) {
      const excelRow = index + 3;
      const numero = String(row[0] ?? '').trim();
      if (!numero || numero === 'Nº Venda') continue;
      const codigo = String(row[4] ?? '').trim();
      const anuncio = findAnuncio.get(codigo);
      const produto = splitProductAndSku(row[3]);
      insertVenda.run(
        numero,
        anuncio?.id || null,
        codigo,
        produto.sku || anuncio?.sku || '',
        produto.nome,
        String(row[9] ?? '').trim(),
        String(row[1] ?? '').trim(),
        excelDate(row[2]),
        Math.trunc(toFloat(row[5])),
        toFloat(row[6]),
        toFloat(row[7]),
        String(row[8] ?? '').trim(),
        link(workbook, 'Vendas', `A${excelRow}`),
      );
      vendasCount += 1;
    }

    db.prepare(`
      INSERT INTO importacoes
      (arquivo_origem, anuncios_importados, promocoes_importadas, vendas_importadas, observacao)
      VALUES (?, ?, ?, ?, ?)
    `).run(
      path.basename(spreadsheetPath),
      anunciosCount,
      promocoesCount,
      vendasCount,
      'Importação via Node/Electron a partir da planilha do Mercado Livre.',
    );

    db.exec('COMMIT');
    console.log(`Importação concluída: anúncios=${anunciosCount}, promoções=${promocoesCount}, vendas=${vendasCount}`);
  } catch (error) {
    db.exec('ROLLBACK');
    throw error;
  } finally {
    db.close();
  }
}

importSpreadsheet();
